src/onehot.jl

```diff
@@ -1,57 +1,286 @@
+const OneHotMatrix{K} = OneHotArray{K, 1}
 
-Base.hcat(x::OneHotVector, xs::OneHotVector...) = OneHotMatrix(length(x), [x, xs...])
+onehotsize(::OneHotArray{K}) where K = Int(K)
 
-batch(xs::AbstractArray{<:OneHotVector}) = OneHotMatrix(length(first(xs)), xs)
+# array interface
+Base.size(oa::OneHotArray{K}) where K = (onehotsize(oa), size(oa.onehots)...)
 
+function Base.getindex(oa::OneHotArray{K, N}, i, is::Vararg{Int, N}) where {K, N}
+  @boundscheck checkbounds(oa, i, is...)
+  oa.onehots[is...][i]
+end
+
+function Base.getindex(oa::OneHotArray{K, N}, i::Colon, is::Vararg{Int, N}) where {K, N}
+  @boundscheck checkbounds(oa, i, is...)
+  oa.onehots[is...]
+end
+
+function Base.getindex(oa::OneHotArray{K}, i::Colon, is...) where {K}
+  @boundscheck checkbounds(oa, i, is...)
+  OneHotArray(oa.onehots[is...])
+end
+
+function Base.getindex(oa::OneHotArray{K}, i::Integer, is::Vararg{Colon, N}) where {K, N}
+  @boundscheck checkbounds(oa, i, is...)
+  map(x->x[i], oa.onehots)
+end
+
+Base.similar(o::OneHotArray, ::Type{T}, dims::Dims{N}) where {T, N} = similar(o.onehots, T, dims)
+
+# printing
+
+function Base.summary(io::IO, oa::OneHotArray)
+  join(io, size(oa), 'x')
+  join(io, [" OneHotArray{", onehotsize(oa), ", ", ndims(oa), ", "])
+  Base.showarg(io, oa.onehots, true)
+  print(io, "}")
+end
+
+# cat
+
+Base.vcat(xss::OneHot{K}...) where K = cat(xss...; dims=Val(1))
+Base.hcat(xss::OneHot{K}...) where K = cat(xss...; dims=Val(2))
+
+function Base.cat(xss::OneHot{K}...; dims) where K
+  isone(::Val{V}) where V = isone(V)
+  isone(v) = Base.isone(v)
+  if isone(dims)
+    @warn "concat OneHot{$K} along dimension 1."
+    Base._cat(Val(1), xss...)
+  else
+    predecessor(::Val{V}) where V = Val(V-1)
+    predecessor(v) = Val(v - 1)
+    yss = reshape(collect(xss), reverse(Base.rdims(predecessor(dims), axes(xss))))
+    OneHotArray(yss)
+  end
+end
+
+Base.vcat(xss::OneHotArray{K}...) where K = cat(xss...; dims=Val(1))
+Base.hcat(xss::OneHotArray{K}...) where K = cat(xss...; dims=Val(2))
+
+function Base.cat(xss::OneHotArray{K}...; dims) where K
+  isone(::Val{V}) where V = isone(V)
+  isone(v) = Base.isone(v)
+  if isone(dims)
+    @warn "concat OneHotArray{$K} along dimension 1."
+    Base._cat(Val(1), xss...)
+  else
+    predecessor(::Val{V}) where V = Val(V-1)
+    predecessor(v) = v - 1
+    sdims = predecessor(dims)
+    xidss = map(xs->xs.onehots, xss)
+    ret = cat(xidss...; dims=sdims)
+    OneHotArray(ret)
+  end
+end
+
+# reshape
+
+"reshape the onehots"
+function ohreshape(parent::OneHotArray{K}, dims) where K
+  onehots = parent.onehots
+  OneHotArray(reshape(onehots, dims))
+end
+
+function Base.reshape(parent::OneHotArray{K}, dims::Dims) where K
+  isequal(prod(dims), length(parent)) || throw(DimensionMismatch("new dimensions $(dims) must be consistent with array size $(length(parent))"))
+  return isequal(K, first(dims)) ?
+    ohreshape(parent, Base.tail(dims)) :
+    Base._reshape(parent, dims)
```

Comment:
also here we shouldn't use julia's internals

Reply:
`Base._reshape` is use to avoid redefine all the reshape dim patterns and handle the reshape of `AbstractArray{Bool}`. It's quite common for defining custom array (like [FillArrays](https://github.com/JuliaArrays/FillArrays.jl/blob/master/src/FillArrays.jl#L214-L225))

Reply:
it's enough to use 
```suggestion
   reshape(parent, dims)
```
then Base will handle that
https://github.com/JuliaLang/julia/blob/0f0bc294f7b18d685c4998ada9c2c33841a74f2d/base/reshapedarray.jl#L112

Reply:
we are overloading it, that would result in infinite recursion.

Reply:
That line is exactly the reason I use `Base._reshape`